### ISSUE TITLE
Permission request code

### ISSUE THREAD

**Issue by NmdOfficial** [mit-cml/appinventor-sources#1520]
I get following IllegalArgumentException with the new permission system:

```
java.lang.IllegalArgumentException: Can only use lower 16 bits for requestCode
        at android.support.v4.app.BaseFragmentActivityApi14.checkForValidRequestCode(BaseFragmentActivityApi14.java:79)
        at android.support.v4.app.FragmentActivity.validateRequestPermissionsRequestCode(FragmentActivity.java:734)
        at android.support.v4.app.ActivityCompat.requestPermissions(ActivityCompat.java:392)
```

Reason in Form.java file:
int nonce = permissionRandom.nextInt(100000);

Solution:
int nonce = permissionRandom.nextInt(65535);

**Comment by ewpatton** [comment]
According to this post <https://stackoverflow.com/questions/25529865/java-lang-illegalargumentexception-can-only-use-lower-16-bits-for-requestcode>, this is due to using FragmentActivity rather than Activity as the base class for Form (or AppInventorCompatActivity if you’re using that). Since we don’t using FragmentActivity, I don’t think we will see this issue in App Inventor, but it would be easy enough to confirm by adding 65536 to the nonce to see if we trigger the exception.

**Comment by NmdOfficial** [comment]
We from Kodular (old name Makeroid) use `Form extends AppCompatActivity` for our forms.
But AI use `public class AppInventorCompatActivity extends Activity`.

Maybe that's the difference.

**Comment by Amerkashi** [comment]
AppCompatActivity extends FragmentActivity which extends Activity

App Inventor Form extends AppInventorCompatActivity which itself extends Activity, so AI won't have the issue:
AI:          int nonce = permissionRandom.nextInt(100000);

AppCompat:
static void checkForValidRequestCode(int requestCode) {
      if ((requestCode & -65536) != 0) {
         throw new IllegalArgumentException("Can only use lower 16 bits for requestCode");
      }
   }

**Comment by ewpatton** [comment]
Ok. So I propose the following. Let's make a constant called `MAX_PERMISSION_NONCE` with a value of 100000 and replace the magic number in the code. This way, if you need to change the value to something different you can, and if we ever touch the code it won't result in a merge conflict since the actual definition is elsewhere.

**Comment by ankritisachan** [comment]
Hi. I want to work on this bug. Please provide me some background reading and how can I reproduce this ?

**Comment by ewpatton** [comment]
@ankritisachan Technically this isn't a bug with App Inventor but rather downstream sources. We generate a request code in the range [0,100000), which works fine for our purposes because of the fact that we extend the Activity class provided by the Android frameworks. However, if one switches to using FragmentActivity as the superclass, it restricts the range to [0,65535]. What I proposed in my earlier comment was that we simply make the upper bound a `static final` field that can be configured on a per-distribution basis to allow for these constraints. We will want to keep the App Inventor value so that it doesn't change the behavior at all. The code in question occurs at [Form.java:2520](https://github.com/mit-cml/appinventor-sources/blob/835e0f954f3b2a8a5c1bf3296634f1a0bcd06c48/appinventor/components/src/com/google/appinventor/components/runtime/Form.java#L2520).

**Comment by ankritisachan** [comment]
Okay, so I understand what needs to be done here. Please correct me if I am wrong.
1) I just need to add a line in Form.java -
public static final int MAX_PERMISSION_NONCE = 100000  and
replace '100000' of line no. 2520 to MAX_PERMISSION_NONCE. 
Am I correct?
2) I read about the background of Fragment class vs FragmentActivity class and when they are used.
I wanted to know if AI is not using FragmentActivity class anytime, why do we need to address this issue in AI? 
How will one 'switch to using FragmentActivity class'(as you pointed out) in AI? Is it possible in AI?
And what is meant by 'distribution' in the above comment? Did you mean 'Android Versions' by it?

**Comment by ankritisachan** [comment]
@ewpatton Need your views on the above comment. Please have a look. Thanks.

**Comment by ewpatton** [comment]
@ankritisachan 

1. Yes, that's correct.
2. There is no way to do this in App Inventor. App Inventor is open source, and others (like @NmdOfficial) have added new functionality and run their own copies. I consider these distributions, similar to how there other distributions of Linux that mix and match new functionality based on user needs (e.g., CentOS, Ubuntu).

While technically it isn't something we need to fix for App Inventor, I think it is also good form for us to not have magic numbers in the code and making them well named constants is a better practice. If we ever decide to extend FragmentActivity in the future, it will also make it easier to migrate. You should make a comment before the constant noting that it must be changed from 100000 to 65535 if we ever do that.

**Comment by ankritisachan** [comment]
@ewpatton Okay, Thanks a lot for clarifying my doubts. I will submit a PR.

**Comment by ewpatton** [comment]
Resolved by 5d0e66cf83541b74180a1c9835d6919b11097142
